plugin-server/src/cdp/async-function-executor.ts

```diff
@@ -0,0 +1,103 @@
+
+        switch (request.asyncFunctionName) {
+            case 'fetch':
+                await this.asyncFunctionFetch(request)
+                break
+            default:
+                status.error('🦔', `[HogExecutor] Unknown async function: ${request.asyncFunctionName}`, loggingContext)
+        }
+
+        return request
+    }
+
+    private async asyncFunctionFetch(request: HogFunctionInvocationAsyncRequest): Promise<any> {
+        // TODO: validate the args
+        const args = request.asyncFunctionArgs ?? []
+        const url: string = args[0]
+        const options = args[1]
+
+        const method = options.method || 'POST'
+        const headers = options.headers || {
+            'Content-Type': 'application/json',
+        }
+        const body = options.body || {}
+
+        const webhook: Webhook = {
+            url,
+            method: method,
+            headers: headers,
+            body: typeof body === 'string' ? body : JSON.stringify(body, undefined, 4),
+        }
+
+        // NOTE: Purposefully disabled for now - once we have callback support we can re-enable
+        // const SPECIAL_CONFIG_ID = -3 // Hardcoded to mean Hog
+        // const success = await this.rustyHook.enqueueIfEnabledForTeam({
+        //     webhook: webhook,
+        //     teamId: hogFunction.team_id,
+        //     pluginId: SPECIAL_CONFIG_ID,
+        //     pluginConfigId: SPECIAL_CONFIG_ID,
+        // })
+
+        const success = false
+
+        // TODO: Temporary test code
+        if (!success) {
+            status.info('🦔', `[HogExecutor] Webhook not sent via rustyhook, sending directly instead`)
+            const response: HogFunctionInvocationAsyncResponse = {
+                ...request,
+            }
+
+            try {
+                const fetchResponse = await trackedFetch(url, {
+                    method: webhook.method,
+                    body: webhook.body,
+                    headers: webhook.headers,
+                    timeout: this.serverConfig.EXTERNAL_REQUEST_TIMEOUT_MS,
+                })
+
+                const maybeJson = await fetchResponse.json().catch(() => null)
+
+                response.vmResponse = {
+                    status: fetchResponse.status,
+                    body: await fetchResponse.text(),
+                    json: maybeJson,
+                }
+            } catch (err) {
+                status.error('🦔', `[HogExecutor] Error during fetch`, { ...request, error: String(err) })
+                response.error = 'Something went wrong with the fetch request.'
+            }
+
+            // NOTE: This feels like overkill but is basically simulating rusty hook's callback that will eventually be implemented
+            await this.kafkaProducer!.produce({
+                topic: KAFKA_CDP_FUNCTION_CALLBACKS,
+                value: Buffer.from(JSON.stringify(response)),
+                key: response.id,
+                waitForAck: true,
```

Comment:
re: waitForAck and future improvements (totally non blocking!), I think we'd benefit from 1) running sync code on all of the incoming Hog function invocations, 2) grouping them by which async function is called (effectively fetch vs sleep?), 3) sending each group into kafka as a full batch, instead of ack-ing per message.

Reply:
I agree - just not getting ahead of ourselves. Given this is all very "test" code anyways.

The actual plan once we have the other service is that this would just return a list of messages to be produced.

We could potentially group them into one kaka message if that works for the downstream service (rustyhook for sure would work given that it is anyways receiving the message and then queueing the individual hooks)